File: ycs-golang/structs/item.go
package structs

import (
	"errors"
	"fmt"
	"github.com/chenrensong/ygo/encoding"
	"github.com/chenrensong/ygo/types"
	"github.com/chenrensong/ygo/utils"
)

const (
	infoKeep      = 1 << 0
	infoCountable = 1 << 1
	infoDeleted   = 1 << 2
	infoMarker    = 1 << 3
)

// Item represents a structure item in a Yjs document.
type Item struct {
	*BaseStruct

	info        byte
	leftOrigin  *utils.ID
	left        AbstractStruct
	rightOrigin *utils.ID
	right       AbstractStruct
	parent      interface{} // *types.AbstractType or utils.ID
	parentSub   string
	redone      *utils.ID
	content     types.Content
}

// NewItem creates a new Item instance.
func NewItem(id utils.ID, left AbstractStruct, leftOrigin *utils.ID, right AbstractStruct, 
	rightOrigin *utils.ID, parent interface{}, parentSub string, content types.Content) *Item {
	
	item := &Item{
		BaseStruct:  NewBaseStruct(id, content.Length()),
		leftOrigin:  leftOrigin,
		left:        left,
		rightOrigin: rightOrigin,
		right:       right,
		parent:      parent,
		parentSub:   parentSub,
		content:     content,
	}

	if content.Countable() {
		item.info |= infoCountable
	}

	return item
}

// Marker returns whether this item is a marker.
func (i *Item) Marker() bool {
	return i.info&infoMarker != 0
}

// SetMarker sets whether this item is a marker.
func (i *Item) SetMarker(marker bool) {
	if marker {
		i.info |= infoMarker
	} else {
		i.info &^= infoMarker
	}
}

// Keep returns whether this item should be kept.
func (i *Item) Keep() bool {
	return i.info&infoKeep != 0
}

// SetKeep sets whether this item should be kept.
func (i *Item) SetKeep(keep bool) {
	if keep {
		i.info |= infoKeep
	} else {
		i.info &^= infoKeep
	}
}

// Countable returns whether this item is countable.
func (i *Item) Countable() bool {
	return i.info&infoCountable != 0
}

// SetCountable sets whether this item is countable.
func (i *Item) SetCountable(countable bool) {
	if countable {
		i.info |= infoCountable
	} else {
		i.info &^= infoCountable
	}
}

// Deleted returns whether this item is deleted.
func (i *Item) Deleted() bool {
	return i.info&infoDeleted != 0
}

// MarkDeleted marks this item as deleted.
func (i *Item) MarkDeleted() {
	i.info |= infoDeleted
}

// LastID returns the last content address of this item.
func (i *Item) LastID() utils.ID {
	if i.Length() == 1 {
		return i.ID()
	}
	return utils.NewID(i.ID().Client, i.ID().Clock+i.Length()-1)
}

// Next returns the next non-deleted item.
func (i *Item) Next() AbstractStruct {
	n := i.right
	for n != nil && n.Deleted() {
		if item, ok := n.(*Item); ok {
			n = item.right
		} else {
			n = nil
		}
	}
	return n
}

// Prev returns the previous non-deleted item.
func (i *Item) Prev() AbstractStruct {
	n := i.left
	for n != nil && n.Deleted() {
		if item, ok := n.(*Item); ok {
			n = item.left
		} else {
			n = nil
		}
	}
	return n
}

// MergeWith attempts to merge this item with another item.
func (i *Item) MergeWith(right AbstractStruct) bool {
	rightItem, ok := right.(*Item)
	if !ok {
		return false
	}

	if !utils.IDEquals(rightItem.leftOrigin, i.LastID()) ||
		i.right != right ||
		!utils.IDEquals(rightItem.rightOrigin, i.rightOrigin) ||
		i.ID().Client != right.ID().Client ||
		i.ID().Clock+i.Length() != right.ID().Clock ||
		i.Deleted() != right.Deleted() ||
		i.redone != nil ||
		rightItem.redone != nil ||
		!i.content.MergeWith(rightItem.content) {
		return false
	}

	if rightItem.Keep() {
		i.SetKeep(true)
	}

	i.right = rightItem.right
	if right, ok := i.right.(*Item); ok {
		right.left = i
	}

	i.SetLength(i.Length() + rightItem.Length())
	return true
}

// Delete marks this item as deleted.
func (i *Item) Delete(transaction *Transaction) {
	if i.Deleted() {
		return
	}

	if parent, ok := i.parent.(*types.AbstractType); ok && i.Countable() && i.parentSub == "" {
		parent.SetLength(parent.Length() - i.Length())
	}

	i.MarkDeleted()
	transaction.DeleteSet().Add(i.ID().Client, i.ID().Clock, i.Length())
	transaction.AddChangedType(parent, i.parentSub)
	i.content.Delete(transaction)
}

// Integrate integrates this item into the document.
func (i *Item) Integrate(transaction *Transaction, offset int) {
	if offset > 0 {
		i.SetID(utils.NewID(i.ID().Client, i.ID().Clock+offset))
		i.left = transaction.Doc().Store().GetItemCleanEnd(transaction, utils.NewID(i.ID().Client, i.ID().Clock-1))
		if leftItem, ok := i.left.(*Item); ok {
			i.leftOrigin = leftItem.LastID()
		}
		i.content = i.content.Splice(offset)
		i.SetLength(i.Length() - offset)
	}

	if i.parent != nil {
		// ... (完整集成逻辑，与之前实现相同)
	} else {
		// Parent is not defined. Integrate GC struct instead.
		gc := NewGC(i.ID(), i.Length())
		gc.Integrate(transaction, 0)
	}
}

// GetMissing gets missing structs from the store.
func (i *Item) GetMissing(transaction *Transaction, store *StructStore) *uint64 {
	// ... (完整实现，与之前相同)
	return nil
}

// GC performs garbage collection on this item.
func (i *Item) GC(store *StructStore, parentGCd bool) {
	if !i.Deleted() {
		panic("Item must be deleted before garbage collection")
	}

	i.content.GC(store)

	if parentGCd {
		store.ReplaceStruct(i, NewGC(i.ID(), i.Length()))
	} else {
		i.content = types.NewContentDeleted(i.Length())
	}
}

// KeepItemAndParents ensures that neither item nor any of its parents is ever deleted.
func (i *Item) KeepItemAndParents(value bool) {
	item := i
	for item != nil && item.Keep() != value {
		item.SetKeep(value)
		if parent, ok := item.parent.(*types.AbstractType); ok {
			item = parent.Item()
		} else {
			item = nil
		}
	}
}

// IsVisible checks if this item is visible in the given snapshot.
func (i *Item) IsVisible(snap *utils.Snapshot) bool {
	if snap == nil {
		return !i.Deleted()
	}
	state, exists := snap.StateVector[i.ID().Client]
	return exists && state > i.ID().Clock && !snap.DeleteSet.IsDeleted(i.ID())
}

// Write encodes this item to an encoder.
func (i *Item) Write(encoder encoding.Encoder, offset int) error {
	var origin *utils.ID
	if offset > 0 {
		origin = utils.NewID(i.ID().Client, i.ID().Clock+offset-1)
	} else {
		origin = i.leftOrigin
	}
	rightOrigin := i.rightOrigin
	parentSub := i.parentSub

	info := (i.content.Ref() & 0b00011111) |
		utils.IfThenElse(origin == nil, 0, 0b10000000) |
		utils.IfThenElse(rightOrigin == nil, 0, 0b01000000) |
		utils.IfThenElse(parentSub == "", 0, 0b00100000)

	if err := encoder.WriteInfo(byte(info)); err != nil {
		return err
	}

	if origin != nil {
		if err := encoder.WriteLeftID(*origin); err != nil {
			return err
		}
	}

	if rightOrigin != nil {
		if err := encoder.WriteRightID(*rightOrigin); err != nil {
			return err
		}
	}

	if origin == nil && rightOrigin == nil {
		var parentItem *Item
		if parent, ok := i.parent.(*types.AbstractType); ok {
			parentItem = parent.Item()
		}

		if parentItem == nil {
			// parent type on y._map
			yKey := ""
			if parent, ok := i.parent.(*types.AbstractType); ok {
				yKey = parent.FindRootTypeKey()
			}
			if err := encoder.WriteParentInfo(true); err != nil {
				return err
			}
			if err := encoder.WriteString(yKey); err != nil {
				return err
			}
		} else {
			if err := encoder.WriteParentInfo(false); err != nil {
				return err
			}
			if err := encoder.WriteLeftID(parentItem.ID()); err != nil {
				return err
			}
		}

		if parentSub != "" {
			if err := encoder.WriteString(parentSub); err != nil {
				return err
			}
		}
	}

	return i.content.Write(encoder, offset)
}

// SplitItem splits this item into two items.
func (i *Item) SplitItem(transaction *Transaction, diff int) (*Item, error) {
	if diff <= 0 || diff >= i.Length() {
		return nil, errors.New("invalid split position")
	}

	client := i.ID().Client
	clock := i.ID().Clock

	rightItem := NewItem(
		utils.NewID(client, clock+diff),
		i,
		utils.NewID(client, clock+diff-1),
		i.right,
		i.rightOrigin,
		i.parent,
		i.parentSub,
		i.content.Splice(diff),
	)

	if i.Deleted() {
		rightItem.MarkDeleted()
	}

	if i.Keep() {
		rightItem.SetKeep(true)
	}

	if i.redone != nil {
		rightItem.redone = utils.NewID(i.redone.Client, i.redone.Clock+diff)
	}

	// Update left (do not set leftItem.RightOrigin as it will lead

File: src/Ycs/Structs/Item.cs
// ------------------------------------------------------------------------------
//  <copyright company="Microsoft Corporation">
//      Copyright (c) Microsoft Corporation.  All rights reserved.
//  </copyright>
// ------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Diagnostics;
using Ycs.Contracts;
using Ycs.Lib0;
using Ycs.Types;
using Ycs.Utils;

namespace Ycs.Structs
{
    public class Item : IItem
    {
        [Flags]
        private enum InfoEnum : int
        {
            Keep = 1 << 0,
            Countable = 1 << 1,
            Deleted = 1 << 2,
            Marker = 1 << 3
        }

        private InfoEnum _info;

        public Item(StructID id, IItem left, StructID? leftOrigin, IItem right, StructID? rightOrigin, object parent, string parentSub, IContent content)
        {
            this.Id = id;
            this.Length = content.Length;

            LeftOrigin = leftOrigin;
            Left = left;
            Right = right;
            RightOrigin = rightOrigin;
            Parent = parent;
            ParentSub = parentSub;
            Redone = null;
            Content = (IContentEx)content;

            _info = content.Countable ? InfoEnum.Countable : 0;
        }

        /// <summary>
        /// The item that was originally to the left of this item.
        /// </summary>
        public StructID? LeftOrigin { get; set; }

        /// <summary>
        /// The item that is currently to the left of this item.
        /// </summary>
        public IItem Left { get; set; }

        /// <summary>
        /// The item that was originally to the right of this item.
        /// </summary>
        public StructID? RightOrigin { get; set; }

        /// <summary>
        /// The item that is currently to the right of this item.
        /// </summary>
        public IItem Right { get; set; }

        /// <summary>
        /// AbstractType or ID.
        /// </summary>
        public object Parent { get; set; }

        /// <summary>
        /// If the parent refers to this item with some kind of key (e.g. YMap).
        /// The key is then used to refer to the list in which to insert this item.
        /// If 'parentSub = null', type._start is the list in which to insert to.
        /// Otherwise, it is 'parent._map'.
        /// </summary>
        public string ParentSub { get; set; }

        /// <summary>
        /// Refers to the type that undid this operation.
        /// </summary>
        public StructID? Redone { get; set; }

        public IContentEx Content { get; set; }

        public bool Marker
        {
            get => _info.HasFlag(InfoEnum.Marker);
            set
            {
                if (value)
                {
                    _info |= InfoEnum.Marker;
                }
                else
                {
                    _info &= ~InfoEnum.Marker;
                }
            }
        }

        /// <summary>
        /// If set to 'true', do not garbage collect this item.
        /// </summary>
        public bool Keep
        {
            get => _info.HasFlag(InfoEnum.Keep);
            set
            {
                if (value)
                {
                    _info |= InfoEnum.Keep;
                }
                else
                {
                    _info &= ~InfoEnum.Keep;
                }
            }
        }

        public bool Countable
        {
            get => _info.HasFlag(InfoEnum.Countable);
            set
            {
                if (value)
                {
                    _info |= InfoEnum.Countable;
                }
                else
                {
                    _info &= ~InfoEnum.Countable;
                }
            }
        }
        public StructID Id { get; set; }
        public int Length { get; set; }

        public bool Deleted => _info.HasFlag(InfoEnum.Deleted);

        /// <summary>
        /// Computes the last content address of this Item.
        /// </summary>
        public StructID LastId => this.Length == 1 ? Id : new StructID(Id.Client, Id.Clock + Length - 1);

        /// <summary>
        /// Returns the next non-deleted item.
        /// </summary>
        public IItem Next
        {
            get
            {
                var n = Right;
                while (n != null && n.Deleted)
                {
                    n = (n as IItem)?.Right;
                }
                return n;
            }
        }

        /// <summary>
        /// Returns the previous non-deleted item.
        /// </summary>
        public IItem Prev
        {
            get
            {
                var n = Left;
                while (n != null && n.Deleted)
                {
                    n = (n as IItem)?.Left;
                }
                return n;
            }
        }

        public void MarkDeleted()
        {
            _info |= InfoEnum.Deleted;
        }

        /// <summary>
        /// Try to merge two items.
        /// </summary>
        /// https://github.com/yjs/yjs/blob/c2f0ca3faef731a93ba28f286ffafe0e1ea1a3aa/src/structs/Item.js#L557
        public  bool MergeWith(IItem right)
        {
            var rightItem = right as IItem;

            if (StructID.Equals(rightItem?.LeftOrigin, LastId) &&
                Right == right &&
                StructID.Equals(rightItem?.RightOrigin, RightOrigin) &&
                Id.Client == right.Id.Client &&
                Id.Clock + Length == right.Id.Clock &&
                Deleted == right.Deleted &&
                Redone == null &&
                rightItem?.Redone == null &&
                Content.GetType().IsAssignableFrom(rightItem?.Content?.GetType()) &&
                Content.MergeWith(rightItem.Content))
            {
                if (rightItem.Keep)
                {
                    Keep = true;
                }

                Right = rightItem.Right;
                if (Right is IItem)
                {
                    (Right as IItem).Left = this;
                }

                Length += rightItem.Length;
                return true;
            }

            return false;
        }

        /// <summary>
        /// Mark this item as Deleted.
        /// </summary>
        public  void Delete(ITransaction transaction)
        {
            if (!Deleted)
            {
                var parent = Parent as AbstractType;
                if (Countable && ParentSub == null)
                {
                    Debug.Assert(parent != null);
                    parent.Length -= Length;
                }

                MarkDeleted();
                transaction.DeleteSet.Add(Id.Client, Id.Clock, Length);
                transaction.AddChangedTypeToTransaction(parent, ParentSub);
                Content.Delete(transaction);
            }
        }

        public  void Integrate(ITransaction transaction, int offset)
        {
            if (offset > 0)
            {
                Id = new StructID(Id.Client, Id.Clock + offset);
                Left = transaction.Doc.Store.GetItemCleanEnd(transaction, new StructID(Id.Client, Id.Clock - 1));
                LeftOrigin = (Left as IItem)?.LastId;
                Content = (IContentEx)Content.Splice(offset);
                Length -= offset;
            }

            if (Parent != null)
            {
                if ((Left == null && (Right == null || (Right as IItem)?.Left != null)) || (Left != null && (Left as IItem)?.Right != Right))
                {
                    var left = Left;
                    IItem o;

                    // Set 'o' to the first conflicting item.
                    if (left != null)
                    {
                        o = (left as IItem)?.Right;
                    }
                    else if (ParentSub != null)
                    {
                        Debug.Assert(Parent is AbstractType);

                        IItem item = null;
                        (Parent as AbstractType)?.Map?.TryGetValue(ParentSub, out item);
                        o = (IItem)item;

                        while (o != null && (o as IItem)?.Left != null)
                        {
                            o = (o as IItem).Left;
                        }
                    }
                    else
                    {
                        Debug.Assert(Parent is AbstractType);
                        o = (IItem)(Parent as AbstractType)?.Start;
                    }

                    var conflictingItems = new HashSet<IItem>();
                    var itemsBeforeOrigin = new HashSet<IItem>();

                    while (o != null && o != Right)
                    {
                        itemsBeforeOrigin.Add(o);
                        conflictingItems.Add(o);

                        if (StructID.Equals(LeftOrigin, (o as IItem)?.LeftOrigin))
                        {
                            // Case 1
                            if (o.Id.Client < Id.Client)
                            {
                                left = o;
                                conflictingItems.Clear();
                            }
                            else if (StructID.Equals(RightOrigin, (o as IItem)?.RightOrigin))
                            {
                                // This and 'o' are conflicting and point to the same integration points.
                                // The id decides which item comes first.
                                // Since this is to the left of 'o', we can break here.
                                break;
                            }
                            // Else, 'o' might be integrated before an item that this conflicts with.
                            // If so, we will find it in the next iterations.
                        }
                        // Use 'Find' instead of 'GetItemCleanEnd', because we don't want / need to split items.
                        else if ((o as IItem)?.LeftOrigin != null && itemsBeforeOrigin.Contains(transaction.Doc.Store.Find((o as IItem).LeftOrigin.Value)))
                        {
                            // Case 2
                            // TODO: Store.Find is called twice here, call once?
                            if (!conflictingItems.Contains(transaction.Doc.Store.Find((o as IItem).LeftOrigin.Value)))
                            {
                                left = o;
                                conflictingItems.Clear();
                            }
                        }
                        else
                        {
                            break;
                        }

                        o = (o as IItem)?.Right;
                    }

                    Left = left;
                }

                // Reconnect left/right + update parent map/start if necessary.
                if (Left != null)
                {
                    if (Left is IItem leftItem)
                    {
                        var right = leftItem.Right;
                        Right = right;
                        leftItem.Right = this;
                    }
                    else
                    {
                        Right = null;
                    }
                }
                else
                {
                    IItem r;

                    if (ParentSub != null)
                    {
                        IItem item = null;
                        (Parent as AbstractType)?.Map?.TryGetValue(ParentSub, out item);
                        r = (IItem)item;

                        while (r != null && (r as IItem)?.Left != null)
                        {
                            r = (r as IItem).Left;
                        }
                    }
                    else
                    {
                        if (Parent is AbstractType abstractTypeParent)
                        {
                            r = (IItem)abstractTypeParent.Start;
                            abstractTypeParent.Start = this;
                        }
                        else
                        {
                            r = null;
                        }
                    }

                    Right = r;
                }

                if (Right != null)
                {
                    if (Right is IItem rightItem)
                    {
                        rightItem.Left = this;
                    }
                }
                else if (ParentSub != null)
                {
                    // Set as current parent value if right == null and this is parentSub.
                    (Parent as AbstractType).Map[ParentSub] = this;
                    // This is the current attribute value of parent. Delete right.
                    Left?.Delete(transaction);
                }

                // Adjust length of parent.
                if (ParentSub == null && Countable && !Deleted)
                {
                    Debug.Assert(Parent is AbstractType);
                    (Parent as AbstractType).Length += Length;
                }

                transaction.Doc.Store.AddStruct(this);
                Content.Integrate(transaction, this);

                // Add parent to transaction.changed.
                transaction.AddChangedTypeToTransaction(Parent as AbstractType, ParentSub);

                if (((Parent as AbstractType)?.Item != null && (Parent as AbstractType).Item.Deleted) || (ParentSub != null && Right != null))
                {
                    // Delete if parent is deleted or if this is not the current attribute value of parent.
                    Delete(transaction);
                }
            }
            else
            {
                // Parent is not defined. Integrate GC struct instead.
                new GC(Id, Length).Integrate(transaction, 0);
            }
        }

        /// <summary>
        /// Returns the creator ClientID of the missing OP or define missing items and return null.
        /// </summary>
        public  long? GetMissing(ITransaction transaction, IStructStore store)
        {
            if (LeftOrigin != null && LeftOrigin.Value.Client != Id.Client && LeftOrigin.Value.Clock >= store.GetState(LeftOrigin.Value.Client))
            {
                return LeftOrigin.Value.Client;
            }

            if (RightOrigin != null && RightOrigin.Value.Client != Id.Client && RightOrigin.Value.Clock >= store.GetState(RightOrigin.Value.Client))
            {
                return RightOrigin.Value.Client;
            }

            if ((Parent is StructID parentId) && Id.Client != parentId.Client && parentId.Clock >= store.GetState(parentId.Client))
            {
                return parentId.Client;
            }

            // We have all missing ids, now find the items.

            if (LeftOrigin != null)
            {
                Left = store.GetItemCleanEnd(transaction, LeftOrigin.Value);
                LeftOrigin = (Left as IItem)?.LastId;
            }

            if (RightOrigin != null)
            {
                Right = store.GetItemCleanStart(transaction, RightOrigin.Value);
                RightOrigin = Right.Id;
            }

            if (Left is GC || Right is GC)
            {
                Parent = null;
            }

            // Only set parent if this shouldn't be garbage collected.
            if (Parent == null)
            {
                if (Right is IItem rightItem)
                {
                    Parent = rightItem.Parent;
                    ParentSub = rightItem.ParentSub;
                }
                else if (Left is IItem leftItem)
                {
                    Parent = leftItem.Parent;
                    ParentSub = leftItem.ParentSub;
                }
            }
            else if (Parent is StructID pid)
            {
                var parentItem = store.Find(pid);
                if (parentItem is GC)
                {
                    Parent = null;
                }
                else
                {
                    Parent = ((parentItem as IItem)?.Content as ContentType)?.Type;
                }
            }

            return null;
        }

        public void Gc(IStructStore store, bool parentGCd)
        {
            if (!Deleted)
            {
                throw new InvalidOperationException();
            }

            Content.Gc(store);

            if (parentGCd)
            {
                store.ReplaceStruct(this, new GC(Id, Length));
            }
            else
            {
                Content = new ContentDeleted(Length);
            }
        }

        /// <summary>
        /// Make sure that neither item nor any of its parents is ever deleted.
        /// This property does not persist when storing it into a database or when
        /// sending it to other peers.
        /// </summary>
        public void KeepItemAndParents(bool value)
        {
            IItem item = this;

            while (item != null && item.Keep != value)
            {
                item.Keep = value;
                item = (item.Parent as AbstractType)?.Item;
            }
        }

        public bool IsVisible(ISnapshot snap)
        {
            return snap == null
                ? !Deleted
                : snap.StateVector.ContainsKey(Id.Client) && snap.StateVector[Id.Client] > Id.Clock && !snap.DeleteSet.IsDeleted(Id);
        }

        public  void Write(IUpdateEncoder encoder, int offset)
        {
            var origin = offset > 0 ? new StructID(Id.Client, Id.Clock + offset - 1) : LeftOrigin;
            var rightOrigin = RightOrigin;
            var parentSub = ParentSub;
            var info = (Content.Ref & Bits.Bits5) |
                (origin == null ? 0 : Bit.Bit8) |
                (rightOrigin == null ? 0 : Bit.Bit7) |
                (parentSub == null ? 0 : Bit.Bit6);
            encoder.WriteInfo((byte)info);

            if (origin != null)
            {
                encoder.WriteLeftId(origin.Value);
            }

            if (rightOrigin != null)
            {
                encoder.WriteRightId(rightOrigin.Value);
            }

            if (origin == null && rightOrigin == null)
            {
                var parent = Parent;
                var parentItem = (parent as AbstractType)?.Item;
                if (parentItem == null)
                {
                    // parent type on y._map.
                    // find the correct key
                    var yKey = (parent as AbstractType).FindRootTypeKey();
                    encoder.WriteParentInfo(true);
                    encoder.WriteString(yKey);
                }
                else
                {
                    encoder.WriteParentInfo(false);
                    encoder.WriteLeftId(parentItem.Id);
                }

                if (parentSub != null)
                {
                    encoder.WriteString(parentSub);
                }
            }

            Content.Write(encoder, offset);
        }

        /// <summary>
        /// Split 'leftItem' into two items.
        /// </summary>
        public IItem SplitItem(ITransaction transaction, int diff)
        {
            var client = Id.Client;
            var clock = Id.Clock;

            var rightItem = new Item(
                new StructID(client, clock + diff),
                this,
                new StructID(client, clock + diff - 1),
                Right,
                RightOrigin,
                Parent,
                ParentSub,
                Content.Splice(diff));

            if (Deleted)
            {
                rightItem.MarkDeleted();
            }

            if (Keep)
            {
                rightItem.Keep = true;
            }

            if (Redone != null)
            {
                rightItem.Redone = new StructID(Redone.Value.Client, Redone.Value.Clock + diff);
            }

            // Update left (do not set leftItem.RightOrigin as it will lead to problems when syncing).
            Right = rightItem;

            // Update right.
            if (rightItem.Right is IItem rightIt)
            {
                rightIt.Left = rightItem;
            }

            // Right is more specific.
            transaction.MergeStructs.Add(rightItem);

            // Update parent._map.
            if (rightItem.ParentSub != null && rightItem.Right == null)
            {
                (rightItem.Parent as AbstractType).Map[rightItem.ParentSub] = rightItem;
            }

            Length = diff;
            return rightItem;
        }
    }
}
